### PR TITLE
Build a single sourmash sketch per genome

### PR DESCRIPTION
**Pull request by peterjc**
Previously was building a double genome sketch, and giving this to the compare command. Now we buid a single sketch per genome, and when we do the compare, give it two sketchs to compare. This means we just need *N* smaller sketches, rather than *N^2* double sketches.

Also now listing A then B (but seems fully symmetric anyway).

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] This change requires a documentation update
- [ ] This is a documentation update

## Action Checklist

- [x] Work on a single issue/concept (if there are multiple separate issues to address, please use a separate pull request for each)
- [ ] Fork the `pyani-plus` repository under your own account (please [allow write access for repository maintainers](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
- [x] Set up an appropriate development environment (please see `CONTRIBUTING.md` for this repository)
- [x] Create a new branch with a short, descriptive name
- [x] Work on this branch
  - [x] style guidelines have been followed (please see `CONTRIBUTING.md` for this repository)
  - [x] new code is commented, especially in hard-to-understand areas
  - [x] corresponding changes to documentation have been made
  - [x] tests for the change have been added that demonstrate the fix or feature works
- [x] Test locally with `pytest-plus -v` **non-passing code will not be merged**
- [x] Rebase against `origin/master`
- [x] Check changes with linting/formatting tools before submission (please see `CONTRIBUTING.md` for this repository)
- [x] Commit branch
- [x] Submit pull request, describing the content and intent of the pull request
- [x] Request a code review
- [x] Continue the discussion at [`Pull requests` section](https://github.com/widdowquinn/pyani-plus/pulls) in the `pyan-plus` repository

### CODE REVIEW

**Comment by peterjc** [comment]
Indicative raw timings with 200 bacteria, nearly 6 mins to build the 200 sig k=51 files with ``sourmash sketch``, and then only about 12s to build the ANI matrix in single call to ``sourmash compare`` (!).

Even with 465 signatures, only needed just under a minute to output an ANI matrix. For 657, almost 2 mins. For 871, about 3.5mins.

We _really_ don't want to be making command line calls to ``sourmash compare`` for each pair, even if sub-second! The overhead may be negligible with slower methods like ANIm or ANIb, but would best be avoided here.

**Comment by peterjc** [comment]
The approach used here for logging the all-vs-all sourmash output to the DB in a single call is via the DB run-id as proposed on #175 and implemented on #176

**Comment by peterjc** [comment]
Note that since the sourmash CLI is written in Python, there is the possibility to call the internals of ``sourmash compare`` directly in order to have a working progress bar. That might be worth considering as a followup PR, but does have downsides/risks.

**Comment by peterjc** [comment]
NOTE: I would prefer to merge #176 before this, and have another branch dealing with the merge conflicts from that: ``sourmash_and_run_id`` which I can use to update this PR.

File: pyani_plus/workflows/snakemake_sourmash.smk [hunk 1]
```diff
@@ -42,32 +42,31 @@ rule sketch:
...
+        run_id=config["run_id"],
         outdir=config["outdir"],
         mode=config["mode"],
         kmersize=config["kmersize"],
         extra=config["extra"],  # This will consist of either `scaled=X` or `num=X`.
     input:
-        "{outdir}/{genomeA}_vs_{genomeB}.sig",
-        genomeA=get_genomeA,
-        genomeB=get_genomeB,
+        expand("{{outdir}}/{genome}.sig", genome=sorted(indir_files)),
```
Comment (widdowquinn) [inline pyani_plus/workflows/snakemake_sourmash.smk:60]: This gets used in the command line as a list of that many filenames… is there a buffer size upper limit that we might encounter here?

Comment (widdowquinn) [reply pyani_plus/workflows/snakemake_sourmash.smk:60]: Maybe write these filenames to a file and use redirection? (does this still potentially hit the limit)?

**Review by widdowquinn — APPROVED** [review]
Incredible speed improvements! Just need to see how divergent the results are ;)

**Comment by peterjc** [comment]
Rebased onto main, after applying the logging changes etc.

**Pull request merged**